[develop] Job level scaling with compute resources and batch launches aggregation

lukeseawalker opened the pull request:
### Description of changes
* Create `InstanceManager` children classes, `NodeListScalingInstanceManager` and `JobLevelScalingInstanceManager`
* `NodeListScalingInstanceManager` hold the previous/current behavior of best effort scaling, which is done considering the whole node list coming from Slurm resume request
* `JobLevelScalingInstanceManager` is the new behavior of all or nothing scaling on a per job basis, which is done considering the Slurm resume file. For now, this new scaling is only applied for exclusive job, i.e. job allocated to nodes that are reserved only for that job execution
* `JobLevelScalingInstanceManager` can be disabled by setting `job_level_scaling` to `False` into resume configuration file
* The instantiation of one of the two children classes of `InstanceManager` is provided by the factory `InstanceManagerFactory`
* The previous resume configuration parameter `max_batch_size` has been split into `launch_max_batch_size` and `update_node_max_batch_size`, to be able to control independently the batch size of launch instances Vs the batch size of assigning the EC2 instances to the Slurm nodes (scontrol, Dynamo and Route53 updates)
* A new resume parameter `terminate_max_batch_size` has been added, to control the batch size of the instances to be terminated
* The new all-or-nothing job-level scaling logic in the `JobLevelScalingInstanceManager`, is able to aggregate the capacity across different compute resource requests or different batch launches, so that the EC2 capacity launched is assigned to the Slurm nodes only if it was possible to start all the capacity. In case it isn't possible to start the whole requested capacity, the logic keep track of the partially launched instances into a `unused_launched_instances` dictionary, and try to assign that capacity to other requests in the same resume call. At the end of the job request loop, the remaining unused capacity is terminated in batches of `terminate_max_batch_size`
* To optimize scaling time and reduce API calls for the `JobLevelScalingInstanceManager` case, capacity for jobs requesting only 1 exclusive node is batched all together, in a single best-effort launch call.
* For the case of `JobLevelScalingInstanceManager`, a new custom logging filter is added to the logger, so that the JobID is tracked down into the log

Resume log example for the submission of the following jobs
```
sbatch --wrap "sleep 100" -N 2 --constraint="[(c5.4xlarge)]" -p q4 --exclusive; 
sbatch --wrap "sleep 100" -N 2 --constraint="[(c5.4xlarge)]" -p q4; 
sbatch --wrap "sleep 100" -N 1 --constraint="[(c5.4xlarge)]" -p q4 --exclusive; 
sbatch --wrap "sleep 100" -N 1 --constraint="[(c5.4xlarge)]" -p q4 --exclusive;
``` 

slurm_resume.log
```
2023-06-28 15:50:41,046 - [slurm_plugin.resume:main] - INFO - ResumeProgram startup.
2023-06-28 15:50:41,047 - [slurm_plugin.resume:_get_config] - INFO - Reading /etc/parallelcluster/slurm_plugin/parallelcluster_slurm_resume.conf
2023-06-28 15:50:41,049 - [slurm_plugin.resume:main] - INFO - ResumeProgram config: SlurmResumeConfig([...])
2023-06-28 15:50:41,049 - [slurm_plugin.resume:_get_slurm_resume] - INFO - Slurm Resume File content: {'jobs': [{'extra': None, 'job_id': 407, 'features': '[(c5.4xlarge)]', 'nodes_alloc': 'q4-dy-c4-1-[10-11]', 'nodes_resume': 'q4-dy-c4-1-[10-11]', 'oversubscribe': 'NO', 'partition': 'q4', 'reservation': None}, {'extra': None, 'job_id': 408, 'features': '[(c5.4xlarge)]', 'nodes_alloc': 'q4-dy-c4-1-[12-13]', 'nodes_resume': 'q4-dy-c4-1-[12-13]', 'oversubscribe': 'OK', 'partition': 'q4', 'reservation': None}, {'extra': None, 'job_id': 409, 'features': '[(c5.4xlarge)]', 'nodes_alloc': 'q4-dy-c4-1-14', 'nodes_resume': 'q4-dy-c4-1-14', 'oversubscribe': 'NO', 'partition': 'q4', 'reservation': None}, {'extra': None, 'job_id': 410, 'features': '[(c5.4xlarge)]', 'nodes_alloc': 'q4-dy-c4-1-15', 'nodes_resume': 'q4-dy-c4-1-15', 'oversubscribe': 'NO', 'partition': 'q4', 'reservation': None}], 'all_nodes_resume': 'q4-dy-c4-1-[10-15]'}
2023-06-28 15:50:41,053 - [slurm_plugin.common:is_clustermgtd_heartbeat_valid] - INFO - Latest heartbeat from clustermgtd: 2023-06-28 15:50:34.266947+00:00
2023-06-28 15:50:41,054 - [slurm_plugin.resume:_resume] - INFO - Launching EC2 instances for the following Slurm nodes: q4-dy-c4-1-[10-15]
2023-06-28 15:50:41,126 - [slurm_plugin.resume:_resume] - INFO - Current state of Slurm nodes to resume: [('q4-dy-c4-1-10', 'ALLOCATED+CLOUD+NOT_RESPONDING+POWERING_UP'), ('q4-dy-c4-1-11', 'ALLOCATED+CLOUD+NOT_RESPONDING+POWERING_UP'), ('q4-dy-c4-1-12', 'MIXED+CLOUD+NOT_RESPONDING+POWERING_UP'), ('q4-dy-c4-1-13', 'MIXED+CLOUD+NOT_RESPONDING+POWERING_UP'), ('q4-dy-c4-1-14', 'ALLOCATED+CLOUD+NOT_RESPONDING+POWERING_UP'), ('q4-dy-c4-1-15', 'ALLOCATED+CLOUD+NOT_RESPONDING+POWERING_UP')]
2023-06-28 15:50:41,152 - [botocore.credentials:load] - INFO - Found credentials from IAM Role: gpu-370-RoleHeadNode-KLLH0UAQ5CSX
2023-06-28 15:50:41,191 - [slurm_plugin.instance_manager:_add_instances_for_nodes] - INFO - Launching instances for Slurm nodes (x2) ['q4-dy-c4-1-14', 'q4-dy-c4-1-15']
2023-06-28 15:50:41,192 - [slurm_plugin.fleet_manager:create_fleet] - INFO - Launching instances with create_fleet API. Parameters: {'LaunchTemplateConfigs': [{'LaunchTemplateSpecification': {'LaunchTemplateName': 'gpu-370-q4-c4-1', 'Version': '$Latest'}, 'Overrides': [{'InstanceType': 'c5.4xlarge', 'SubnetId': 'subnet-0b48ed99988e56110'}]}], 'TargetCapacitySpecification': {'TotalTargetCapacity': 2, 'DefaultTargetCapacityType': 'on-demand'}, 'Type': 'instant', 'OnDemandOptions': {'AllocationStrategy': 'lowest-price', 'SingleInstanceType': True, 'SingleAvailabilityZone': True, 'MinTargetCapacity': 1, 'CapacityReservationOptions': {'UsageStrategy': 'use-capacity-reservations-first'}}}
2023-06-28 15:50:48,130 - [slurm_plugin.instance_manager:_update_slurm_node_addrs_and_failed_nodes] - INFO - Nodes are now configured with instances: (x2) ["('q4-dy-c4-1-14', EC2Instance(id='i-01ff7c213128a2379', private_ip='192.168.111.115', hostname='ip-192-168-111-115', launch_time=datetime.datetime(2023, 6, 28, 15, 50, 42, tzinfo=tzlocal()), slurm_node=None))", "('q4-dy-c4-1-15', EC2Instance(id='i-0739d1e42597aa4cd', private_ip='192.168.105.66', hostname='ip-192-168-105-66', launch_time=datetime.datetime(2023, 6, 28, 15, 50, 42, tzinfo=tzlocal()), slurm_node=None))"]
2023-06-28 15:50:48,130 - [slurm_plugin.instance_manager:_store_assigned_hostnames] - INFO - Saving assigned hostnames in DynamoDB
2023-06-28 15:50:48,164 - [slurm_plugin.instance_manager:_store_assigned_hostnames] - INFO - Database update: COMPLETED
2023-06-28 15:50:48,165 - [slurm_plugin.instance_manager:_update_dns_hostnames] - INFO - Updating DNS records for Z09410041GY3CM0CGB9WP - gpu-370.pcluster.
2023-06-28 15:50:48,420 - [slurm_plugin.instance_manager:_update_dns_hostnames] - INFO - DNS records update: COMPLETED
2023-06-28 15:50:48,421 - [slurm_plugin.instance_manager:_add_instances_for_job] - INFO - JobID 407 - Adding instances for nodes (x2) ['q4-dy-c4-1-10', 'q4-dy-c4-1-11']
2023-06-28 15:50:48,421 - [slurm_plugin.instance_manager:_launch_instances_for_job] - INFO - JobID 407 - Launching instances for nodes (x2) ['q4-dy-c4-1-10', 'q4-dy-c4-1-11']
2023-06-28 15:50:48,422 - [slurm_plugin.fleet_manager:create_fleet] - INFO - JobID 407 - Launching instances with create_fleet API. Parameters: {'LaunchTemplateConfigs': [{'LaunchTemplateSpecification': {'LaunchTemplateName': 'gpu-370-q4-c4-1', 'Version': '$Latest'}, 'Overrides': [{'InstanceType': 'c5.4xlarge', 'SubnetId': 'subnet-0b48ed99988e56110'}]}], 'TargetCapacitySpecification': {'TotalTargetCapacity': 2, 'DefaultTargetCapacityType': 'on-demand'}, 'Type': 'instant', 'OnDemandOptions': {'AllocationStrategy': 'lowest-price', 'SingleInstanceType': True, 'SingleAvailabilityZone': True, 'MinTargetCapacity': 2, 'CapacityReservationOptions': {'UsageStrategy': 'use-capacity-reservations-first'}}}
2023-06-28 15:50:55,341 - [slurm_plugin.instance_manager:_add_instances_for_job] - INFO - JobID 407 - Successful launched all instances for nodes (x2) ['q4-dy-c4-1-10', 'q4-dy-c4-1-11']
2023-06-28 15:50:55,359 - [slurm_plugin.instance_manager:_update_slurm_node_addrs] - INFO - JobID 407 - Nodes are now configured with instances: (x2) ["('q4-dy-c4-1-10', EC2Instance(id='i-0eace2f1ad4f8ea3b', private_ip='192.168.107.146', hostname='ip-192-168-107-146', launch_time=datetime.datetime(2023, 6, 28, 15, 50, 49, tzinfo=tzlocal()), slurm_node=None))", "('q4-dy-c4-1-11', EC2Instance(id='i-0863d361f6a82d432', private_ip='192.168.104.221', hostname='ip-192-168-104-221', launch_time=datetime.datetime(2023, 6, 28, 15, 50, 49, tzinfo=tzlocal()), slurm_node=None))"]
2023-06-28 15:50:55,359 - [slurm_plugin.instance_manager:_store_assigned_hostnames] - INFO - JobID 407 - Saving assigned hostnames in DynamoDB
2023-06-28 15:50:55,371 - [slurm_plugin.instance_manager:_store_assigned_hostnames] - INFO - JobID 407 - Database update: COMPLETED
2023-06-28 15:50:55,372 - [slurm_plugin.instance_manager:_update_dns_hostnames] - INFO - JobID 407 - Updating DNS records for Z09410041GY3CM0CGB9WP - gpu-370.pcluster.
2023-06-28 15:50:55,725 - [slurm_plugin.instance_manager:_update_dns_hostnames] - INFO - JobID 407 - DNS records update: COMPLETED
2023-06-28 15:50:55,727 - [slurm_plugin.instance_manager:_add_instances_for_nodes] - INFO - Launching instances for Slurm nodes (x2) ['q4-dy-c4-1-12', 'q4-dy-c4-1-13']
2023-06-28 15:50:55,727 - [slurm_plugin.fleet_manager:create_fleet] - INFO - Launching instances with create_fleet API. Parameters: {'LaunchTemplateConfigs': [{'LaunchTemplateSpecification': {'LaunchTemplateName': 'gpu-370-q4-c4-1', 'Version': '$Latest'}, 'Overrides': [{'InstanceType': 'c5.4xlarge', 'SubnetId': 'subnet-0b48ed99988e56110'}]}], 'TargetCapacitySpecification': {'TotalTargetCapacity': 2, 'DefaultTargetCapacityType': 'on-demand'}, 'Type': 'instant', 'OnDemandOptions': {'AllocationStrategy': 'lowest-price', 'SingleInstanceType': True, 'SingleAvailabilityZone': True, 'MinTargetCapacity': 1, 'CapacityReservationOptions': {'UsageStrategy': 'use-capacity-reservations-first'}}}
2023-06-28 15:51:02,762 - [slurm_plugin.instance_manager:_update_slurm_node_addrs_and_failed_nodes] - INFO - Nodes are now configured with instances: (x2) ["('q4-dy-c4-1-12', EC2Instance(id='i-08a6c8e16a4e3fe19', private_ip='192.168.110.23', hostname='ip-192-168-110-23', launch_time=datetime.datetime(2023, 6, 28, 15, 50, 57, tzinfo=tzlocal()), slurm_node=None))", "('q4-dy-c4-1-13', EC2Instance(id='i-0a41ecaec10401b12', private_ip='192.168.104.165', hostname='ip-192-168-104-165', launch_time=datetime.datetime(2023, 6, 28, 15, 50, 57, tzinfo=tzlocal()), slurm_node=None))"]
2023-06-28 15:51:02,763 - [slurm_plugin.instance_manager:_store_assigned_hostnames] - INFO - Saving assigned hostnames in DynamoDB
2023-06-28 15:51:02,770 - [slurm_plugin.instance_manager:_store_assigned_hostnames] - INFO - Database update: COMPLETED
2023-06-28 15:51:02,771 - [slurm_plugin.instance_manager:_update_dns_hostnames] - INFO - Updating DNS records for Z09410041GY3CM0CGB9WP - gpu-370.pcluster.
2023-06-28 15:51:02,961 - [slurm_plugin.instance_manager:_update_dns_hostnames] - INFO - DNS records update: COMPLETED
2023-06-28 15:51:02,962 - [slurm_plugin.resume:_resume] - INFO - Successfully launched nodes (x6) ['q4-dy-c4-1-10', 'q4-dy-c4-1-11', 'q4-dy-c4-1-12', 'q4-dy-c4-1-13', 'q4-dy-c4-1-14', 'q4-dy-c4-1-15']
2023-06-28 15:51:02,964 - [slurm_plugin.resume:main] - INFO - ResumeProgram finished.
```


Signed-off-by: Luca Carrogu <carrogu@amazon.com>

### Tests
* added unit tests for all the new code added
* manually tested

### References
* Link to related PRs in other packages (i.e. cookbook, node).
* Link to documentation useful to understand the changes.

### Checklist
- [X] Make sure you are pointing to **the right branch** and add a label in the PR title (i.e. **2.x** vs **3.x**)
- [X] Check all commits' messages are clear, describing what and why vs how.
- [X] Make sure **to have added unit tests or integration tests** to cover the new/modified code.
- [X] Check if documentation is impacted by this change.

Please review the [guidelines for contributing](../CONTRIBUTING.md) and [Pull Request Instructions](https://github.com/aws/aws-parallelcluster/wiki/Git-Pull-Request-Instructions).

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

NSsirena reviewed (commented):
A general comment over this PR

In this package we have the tendency to create very convoluted functions that are complex to read and maintain.

Also we have the tendency to favor inheritance over composability.
This also makes the code more rigid and hard to adapt and evolve.

These are out-of-scope for this PR but we should reserve some time to review the infrastructure and design of this package.

lukeseawalker committed:
Preserve exception message if exception to raise is same of actual exception

Signed-off-by: Luca Carrogu <carrogu@amazon.com>

lukeseawalker committed:
Remove unused import

Signed-off-by: Luca Carrogu <carrogu@amazon.com>

lukeseawalker committed:
Split `InstanceManager` into `NodeListScalingInstanceManager` and `JobLevelScalingInstanceManager`

Create `InstanceManager children classes, `NodeListScalingInstanceManager` and `JobLevelScalingInstanceManager`

* `NodeListScalingInstanceManager` hold the previous/current behavior of best effort scaling, which is done considering the whole node list coming from Slurm resume request
* `JobLevelScalingInstanceManager` is the new behavior of all or nothing scaling on a per job basis, which is done considering the Slurm resume file. For now, this new scaling is only applied for exclusive job, i.e. job allocated to nodes that are reserved only for that job execution
* `JobLevelScalingInstanceManager` can be disabled by setting `job_level_scaling` to `False` into resume configuration file
* The two children classes of InstanceManager are provided by the factory `InstanceManagerFactory`
* The new all-or-nothing job-level scaling logic in the `JobLevelScalingInstanceManager`, is able to aggregate the capacity across different compute resource requests or different batch launches, so that the EC2 capacity launched is assigned to the Slurm nodes only if it was possible to start all the capacity. In case it isn't possible to start the whole requested capacity, the logic keep track of the partially launched instances into a `unused_launched_instances` dictionary, and try to assign that capacity to other requests in the same resume call. At the end of the job request loop, the remaining unused capacity is terminated in batches of `terminate_max_batch_size`

Signed-off-by: Luca Carrogu <carrogu@amazon.com>

lukeseawalker committed:
Split `max_batch_size` into `launch_max_batch_size` and `update_node_max_batch_size`

The previous resume configuration parameter `max_batch_size` has been split into `launch_max_batch_size` and `update_node_max_batch_size`, to be able to control independently the batch size of launch instances Vs the batch size of assigning the EC2 instances to the Slurm nodes (scontrol, Dynamo and Route53 updates)

A new resume parameter `terminate_max_batch_size` has been added, to control the batch size of the instances to be terminated

Signed-off-by: Luca Carrogu <carrogu@amazon.com>

lukeseawalker committed:
Optimize job level scaling for single node jobs

To optimize scaling time and reduce API calls for the `JobLevelScalingInstanceManager` case, capacity for jobs requesting only 1 exclusive node is batched all together, in a single best-effort launch call.

Signed-off-by: Luca Carrogu <carrogu@amazon.com>

lukeseawalker committed:
Add custom logging filter to add JobID info

For the case of `JobLevelScalingInstanceManager`, a new custom logging filter is added to the logger, so that the JobID is tracked down into the log

Signed-off-by: Luca Carrogu <carrogu@amazon.com>

lukeseawalker committed:
Unit tests covering new job level scaling logic

Signed-off-by: Luca Carrogu <carrogu@amazon.com>

EddyMM commented on src/slurm_plugin/common.py:
```diff
@@ -73,7 +73,10 @@ def wrapper(*args, **kwargs):
+                        if isinstance(e, exception_to_raise):
+                            raise e  # preserve exception message if exception to raise is same of actual exception
+                        else:
+                            raise exception_to_raise
```
NIT: `raise e if isinstance(exception_to_raise, e) else exception_to_raise`
(Just to reduce the nesting)

EddyMM commented on src/slurm_plugin/instance_manager.py:
```diff
@@ -87,49 +145,42 @@ def __init__(
             resource_name, region_name=region, config=boto3_config
         )
+        self.failed_nodes = {}
+        self.unused_launched_instances = {}
```
Unused launched instances apply to Job Level Scaling only. Can we move `unused_launched_instances` and `_clear_unused_launched_instances` to the `JobLevelScalingInstancemanager` to avoid a single child class relying on the parent's variables/method?

lukeseawalker replied:
abs yes

EddyMM commented on src/slurm_plugin/instance_manager.py:
```diff
@@ -619,3 +566,576 @@ def _create_request_for_nodes(table_name, node_names):
+            )
+            self._update_failed_nodes(set(nodes_difference), "InvalidNodenameError")
+        return SlurmResumeData(
+            nodes_oversubscribe=list(dict.fromkeys(nodes_oversubscribe)),
```
Why use `dict.fromkeys` instead of `set`? If I understand correctly, `dict.fromkeys` preserves order at the cost of a performance difference with sets but we're not relying on order preservation, are we?

lukeseawalker replied:
I wanted to preserve the order of the nodes_oversubscribe list

EddyMM commented on src/slurm_plugin/instance_manager.py:
```diff
@@ -619,3 +566,576 @@ def _create_request_for_nodes(table_name, node_names):
+                    for batch_nodes in grouper(slurm_node_list, launch_batch_size):
+                        try:
+                            launched_ec2_instances = fleet_manager.launch_ec2_instances(
+                                len(batch_nodes), job_id=job.job_id
```
Can we set the job id filter at this level? (Instead of changing the `launch_ec2_instances` function parameters [job_id] which are shared with node list scaling)

lukeseawalker replied:
I couldn't find a good way, because fleet_manager instantiate it's own logger. I could have returned the fleet_manager logger object when calling the get_manager, but I didn't like the approach.

For the case of node list scaling, the job_id will be default to None, and the custom logger isn't added in that case.

EddyMM commented on src/slurm_plugin/instance_manager.py:
```diff
@@ -619,3 +566,576 @@ def _create_request_for_nodes(table_name, node_names):
+                len(instance_list)
+                for compute_resources in instances_launched.values()
+                for instance_list in compute_resources.values()
+            ) == len(job.nodes_resume):
```
Should we consider using the sum of nodes in `nodes_to_launch` instead of `job.nodes_resume`? (In case some nodes are not actually used in the launch e.g. they have an invalid name - `_parse_requested_instances`)

lukeseawalker replied:
good point

EddyMM commented on src/slurm_plugin/instance_manager.py:
```diff
@@ -619,3 +566,576 @@ def _create_request_for_nodes(table_name, node_names):
+                    # setting the nodes into DOWN.
+                    # EC2 instances not yet assigned, are going to fail during bootstrap,
+                    # because no entry in the DynamoDB table would be found
+                    self._update_failed_nodes(set(job.nodes_resume))
```
Same case here: `nodes_to_launch` might be the preferred list to use.

EddyMM commented on src/slurm_plugin/instance_manager.py:
```diff
@@ -619,3 +566,576 @@ def _create_request_for_nodes(table_name, node_names):
+                try:
+                    logger.info(
+                        "Successful launched all instances for nodes %s",
+                        print_with_count(job.nodes_resume),
```
`nodes_to_launch` instead of `job.nodes_resume`?

EddyMM commented on src/slurm_plugin/instance_manager.py:
```diff
@@ -619,3 +566,576 @@ def _create_request_for_nodes(table_name, node_names):
+                    ),
+                )
+                self.unused_launched_instances |= instances_launched
+                self._update_failed_nodes(set(job.nodes_resume), "LimitedInstanceCapacity")
```
Should we consider centralising the error code names (`LimitedInstanceCapacity` and `InsufficientInstanceCapacity`)?

lukeseawalker replied:
I would consider it in another PR

lukeseawalker committed:
Consider only successful parsed node names in scaling logic

Signed-off-by: Luca Carrogu <carrogu@amazon.com>

lukeseawalker committed:
Fix unit tests for job-level scaling

Signed-off-by: Luca Carrogu <carrogu@amazon.com>

EddyMM commented on src/slurm_plugin/resume.py:
```diff
@@ -75,8 +77,14 @@ def _get_config(self, config_file_path):
+        self.launch_max_batch_size = config.getint(
+            "slurm_resume", "launch_max_batch_size", fallback=self.DEFAULTS.get("launch_max_batch_size")
+        )
+        self.update_node_max_batch_size = config.getint(
```
NIT: This is related to the Route53 limit of 500? The name might be confused with something else e.g. rate of updating the node state in Slurm etc.
Could we use something else like `node_info_upload_max_batch_size`

lukeseawalker replied:
this is used when EC2 instance is assigned to a Slurm node, and we call downstream scontrol update, update dynamo, update route53 (if enabled). This is why I have been calling it "update"

EddyMM commented on tests/slurm_plugin/test_instance_manager.py:
```diff
@@ -66,1196 +68,604 @@ def instance_manager(self, mocker):
-        mocker,
-        instance_manager,
+    def test_delete_instances(
+        self, boto3_stubber, instance_ids_to_name, batch_size, mocked_boto3_request, instance_manager, job_level_scaling
```
NIT: I think `instance_ids_to_name` should be `instance_ids_to_terminate`.

EddyMM replied:
Also, why does this test have a `job_level_scaling` parameter? I haven't seen its usages in the function body.
Same case for the `test_get_cluster_instances`, `test_get_console_output` tests.

lukeseawalker replied:
> why does this test have a job_level_scaling parameter

sorry, not needed. It was a leftover when I started the implementation and the two InstanceManger children classes didn't exist yet. I'm removing it

lukeseawalker replied:
it's used by the instance_manager fixture. I will remove it

EddyMM commented on tests/slurm_plugin/test_resume.py:
```diff
@@ -367,17 +376,289 @@ def test_resume_config(config_file, expected_attributes, test_datadir, mocker):
     ],
 )
-def test_resume_launch(
+def test_resume_node_list_scaling(
```
This test covers both node-list and job-level scaling. Could we keep the name as `test_resume_launch` or `test_resume_scaling`?

EddyMM commented on CHANGELOG.md:
```diff
@@ -9,7 +9,7 @@ This file is used to list changes made in each version of the aws-parallelcluste
 
 **CHANGES**
-- Perform job level scaling for exclusive jobs, by reading at job information from `SLURM_RESUME_FILE`. 
+- Perform all-or-nothing / job-level scaling for exclusive jobs, by reading at job information from `SLURM_RESUME_FILE`. 
```
... by reading ~at~ job information from `SLURM_RESUME_FILE`.
